### ISSUE TITLE
Compliance with Rule 2.5

### ISSUE THREAD

**Issue by pavelrappo** [reactive-streams/reactive-streams-jvm#410]
> A `Subscriber` MUST call `Subscription.cancel()` on the given `Subscription` after an `onSubscribe` signal if it already has an active `Subscription`.

I think it's worth pointing out in examples (and may be even producing a TCK test), that subscribers must use appropriate synchronization in `onSubscribe`. Even though there is rule 1.3:

> `onSubscribe`, `onNext`, `onError` and `onComplete` signaled to a `Subscriber` MUST be signaled in a thread-safe manner—and if performed by multiple threads—use external synchronization.

multiple `Publishers` may not coordinate their actions for a subscriber, thus `onSubscribe` may be called concurrently even though each of the publishers call it in a thread-safe manner.

For example, synchronous subscriber could use something like this:

```java
            private final AtomicBoolean hasSubscription = new AtomicBoolean();
            private Subscription subscription;
            
            @Override 
            public void onSubscribe(Subscription s) {
                Objects.requireNonNull(s);
                if (!hasSubscription.compareAndSet(false, true)) {
                    s.cancel();
                } else {
                    subscription = s;
                    // ...
                }
            }
```

**Comment by ktoso** [comment]
Good idea, indeed a Subscriber must be defensive in its onSubscribe implementation, even though for each of the individual pairs we're "safe" due to the 1.3 rule that we impose onto Publishers.

Our test for 2.5 covers this but in a sequential simple setting, we can add one to issue the onSubscribes concurrently indeed.

**Comment by pavelrappo** [comment]
@ktoso I also found a case (even though an extremely unlikely one) which permits an incoherent calls to a `Subscriber`'s methods should a `Publisher` attempt to subscribe this subscriber after the subscriber has been already subscribed to another publisher. Even if calls to this subscriber's methods happen in a thread-safe manner across all publishers.

Consider that even if a `Subscriber` calls `Subscription.cancel` in `onSubscribe` after identifying it already has an active `Subscription`, the `Publisher` can still deliver `onComplete` or `onError` due to a possible propagation delay ([rule 1.8](https://github.com/reactive-streams/reactive-streams-jvm#1.8)). Worst-case scenario -- an unrelated call to `onComplete` or `onError` potentially leading to  an arbitrary size of mess in processing.

Unfortunately in this case not much can be done, but hoping `Subscription.cancel` is responsive enough to terminate the subscription immediately.

**Comment by viktorklang** [comment]
The "only" reason for requiring cancellation if multiple subscriptions are
detected is to fail-fast (throwing is not an option). Since it is only
there for dealing with spec violations, it need not impose performance
penalty for conformant implementations. Mandating onSubscribe to deal with
concurrency control wouldn't help either because onComplete and onError can
be sent without a preceding request-call.

-- 
Cheers,
√

**Comment by ktoso** [comment]
Hmm, interesting point. Our implementation is safe since we linearize the signals via the actor mailbox, so thought it would not be too imposing, but perhaps it indeed is hm.

Is it really just a "fail fast" though? Either it cancels and notices, or what... "eventually notices"? That sounds a bit unrealistic - either you stored that subscription somewhere, or you didn't - so if you did, you may have replaced the existing one (illegal), or dropped the new one on the floor (resulting in a "hanging" subscription). Should we at least do some recommendation for these cases?

**Comment by viktorklang** [comment]
There is already a recommendation though: cancel subscriptions you receive
beyond the first one, because that condition signifies that someone is
violating the spec. (In the case of spec violations there cannot really be
any guarantees anymore)

-- 
Cheers,
√

**Comment by akarnokd** [comment]
There are cases where the same `Subscriber` instance can be reused to handle multiple `Publisher`s, such as concatenating them one after the other. This should be legal when the reuse happens in response to an `onError` or `onComplete`, according to the `has an active Subscription` clause, because the subscription of the previous `Publisher` should be considered cancelled - not active - in those method calls anyway.

Sometimes, *external synchronization* can be best ensured inside the `Publisher`, sometimes in the `Subscriber`.

If you think in a library, you can localize the penalties of the spec into a wrapper-`Subscriber` and have an "inner platform" that can run on different protocols or relaxed rules. See [StrictSubscriber](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/subscribers/StrictSubscriber.java#L25) we had introduced in RxJava 2 to cope with the overhead and implications of a couple of rules.

**Comment by pavelrappo** [comment]
> There is already a recommendation though: cancel subscriptions you receive
beyond the first one, because that condition signifies that someone is
violating the spec. (In the case of spec violations there cannot really be
any guarantees anymore)

How would one reliably ensure a subscription is not the first one?

**Comment by viktorklang** [comment]
If you can't tell, then that's fine.

-- 
Cheers,
√

On Nov 9, 2017 19:40, "pavelrappo" <notifications@github.com> wrote:

There is already a recommendation though: cancel subscriptions you receive
beyond the first one, because that condition signifies that someone is
violating the spec. (In the case of spec violations there cannot really be
any guarantees anymore)

How would one reliably ensure a subscription is not the first one?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/reactive-streams/reactive-streams-jvm/issues/410#issuecomment-343250300>,
or mute the thread

**Comment by viktorklang** [comment]
Is everything addressed here?

**Comment by pavelrappo** [comment]
> Is everything addressed here?

@viktorklang, I believe so. Thanks.

**Comment by thekalinga** [comment]
@viktorklang 
> If you can't tell, then that's fine.

I did not fully understand this. If this is OK, why do we need this rule?

**Comment by viktorklang** [comment]
@thekalinga We have it to reduce the risk of erronous usages. Subscriber is only valid for one subscription in the general case.

**Comment by thekalinga** [comment]
In which case, the subscriber must have a way to determine this, which will contradict "If you can't tell, then that's fine."

**Comment by viktorklang** [comment]
@thekalinga We don't want all implementations to pay the price of doing coordination to detect spec-violating usages. Think about the interaction patterns which are possible *for all erronous usages*. This is about failing fast if one detects erronous states.

**Comment by thekalinga** [comment]
In which case, can the be changed to a `RECOMMENDED` rather than `MUST`?

**Comment by thekalinga** [comment]
> A Subscriber is RECOMMENDED to call Subscription.cancel() on the given Subscription after an onSubscribe signal if it already has an active Subscription.

**Comment by viktorklang** [comment]
@thekalinga SHOULD is probably the better verbiage, but I suspect a clarification in the Intents section of that rule is good enough in this case. Would you agree?

**Comment by thekalinga** [comment]
Sure. Thanks
